### PR TITLE
Implement system-wide authentication 

### LINKED ISSUE #502: PART 1 - JWT security
## User story
As a administrator or instructor user,
I want to know that students can not copy urls and get around security

## Description
Prevent students from copying a URL to gain access to the EdEHR

## Expected behaviour
Only the user who clicked the link in the LMS can be the user in the EdEHR

## Technical 

Server API
- return a JWT after successful LTI post (see lti controller)
- get the client side working and then
- don't include parameters in the query string. Put them in the JWT.

Client
- on load look for the JWT and extract the information currently taken from the query string.
- store this information as is done now.

At this point the system no longer passes sensitive information in the URL.

See PART 2

### LINKED ISSUE #503: PART 2 - JWT
See issue #502 for Part 1
Part 2 uses the JWT to authenticate the user when making API calls to the server.

Client
place the JWT into all headers returned from the client.

Server
- create middleware to look for and extract JWT.
- use this middleware on all routes that support user actions
- throw 403 if not auth'd

### PR DESCRIPTION
**Pull request by viniciusarre**
Closes #502 and closes #503

### CODE REVIEW

**Comment by viniciusarre** [comment]
@bryan-gilbert A few comments on this. I have noticed that, in the issues, it's not mentioned the roles which need to be created for other users. 

In our call last week, we've mentioned a potential 403 response for the cases where a user is unauthorised (but still authenticated) to access the system. I think that, in order to implement this properly, we'd need to address several potential roles. Like instructor, student and (as mentioned in issue #275) admin. 

What do you think about this? Would this be a necessary protection layer? I think it could be quite easy to accomplish. 

I look forward to reading your thoughts!

Thank you.

File: server/src/mcr/auth/auth-controller.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    if (debug) console.log('_getTokenContent')
+    if(req.headers.authorization) {
+      try {
+        const result = this.authenticate(req.headers.authorization)
+        res.status(200).json(result)
+      } catch(err) {
+        res.status(401).send(err)
+      }
+    } else {
+      res.status(401).send('AuthController._getTokenContent --- Token is required!')
```
Comment (bryan-gilbert) [inline server/src/mcr/auth/auth-controller.js:57]: This message will be seen by end users. So make it simple and say "Token is required!"

File: server/src/mcr/auth/auth-controller.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      } catch(err) {
+        res.status(401).send(`AuthController -- Invalid token ${err}`)
+      }
+    } else {
+      if(debug) console.log('Token is required!', req.body)
+      res.status(401).send('Token is required!')
+    }
+  }
+
+  _getTokenContent (req, res) {
```
Comment (bryan-gilbert) [inline server/src/mcr/auth/auth-controller.js:47]: Add /** docs and say this handles X end point

File: server/src/mcr/auth/auth-controller.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    return jwt.sign(data, TOKEN_SECRET)
+  }
+
+  createRefreshToken (token) {
+    if(debug) console.log('createRefreshToken')
+    //set to expire in 1 minute
+    return jwt.sign({token}, TOKEN_SECRET, { expiresIn: '1m'})
+  }
+
+  _getAuthToken (req, res) {
```
Comment (bryan-gilbert) [inline server/src/mcr/auth/auth-controller.js:29]: Add /** docs and say this handles X end point

File: server/src/mcr/auth/auth-controller.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+const jwt = require('jsonwebtoken')
+const debug = false
+
+
+export default class AuthController {
+  constructor () {
+    if(debug) console.log('authController')
+    
+  }
+  authenticate (token) {
```
Comment (bryan-gilbert) [inline server/src/mcr/auth/auth-controller.js:12]: Add /** docs and say what is returned and what are the expected exceptions

File: client/src/App.vue [hunk 5]
```diff
@@ -22,19 +23,54 @@ export default {
...
+          } else {
+            visitId = payload.visitId
+          }
+        })
+        .then(() => {
           return this._loadApiUrl(apiUrl)
         })
         .then(() => {
-          let visitId = params2['visit']
+          // let visitId = params2['visit'] || visitId
```
Comment (bryan-gilbert) [inline client/src/App.vue:73]: Revisit the following and see if it is needed. The visit id exists as of line 66 or else user doesn't get here. So remove the further checks

**Pull request merged**